wasi/wasi.go

```diff
@@ -107,6 +110,48 @@ func Stderr(writer io.Writer) Option {
 	}
 }
 
+// wasiStringArray holds null-terminated strings. It ensures that
+// its length and total buffer size don't exceed the max of uint32.
+type wasiStringArray struct {
+	strings      [][]byte
+	totalBufSize uint32
+}
+
+// newWASIStringArray creates a cStringArray from the given string slice. It returns an error
+// if the length or the total buffer size of the result WASIStringArray exceeds the max of uint32
+func newWASIStringArray(args []string) (*wasiStringArray, error) {
+	if args == nil {
+		return &wasiStringArray{strings: [][]byte{}}, nil
+	}
+	if len(args) > math.MaxUint32 {
+		return nil, fmt.Errorf("the length of the args exceeds the max of uint32: %v", len(args))
+	}
+	strings := make([][]byte, len(args))
+	totalBufSize := uint32(0)
+	for i, arg := range args {
+		argLen := uint64(len(arg)) + 1 // + 1 for '\x00'
```

Comment:
maybe add some unicode to the tests so that this breaks?

Reply:
Added. Thanks.